### PR TITLE
Replace the most critical unordered_maps and maps with faster maps.

### PR DESCRIPTION
**Pull request by hrydgard**
Implemented two new hash maps, FastHashMap and PrehashMap. 

In microbenchmarks on 20-byte keys, FastHashMap beats unordered_map by 10-50% depending on workload, and beats map by 400%.

Though of course overall this is just another 1-2% of performance, with enough of those real change starts to happen.

The removal strategy simply uses a flag to avoid false negatives (see Deletion here: https://en.wikipedia.org/wiki/Linear_probing). This does affect the performance of failed lookups in tables that we don't clear often, but those are not too common. Still, might be worth trying to improve that.

### CODE REVIEW

**Comment by hrydgard** [comment]
Should probably fix the removal strategy before we merge this.

**Comment by hrydgard** [comment]
Decided on a different strategy, rebuild the map if there's too many removed-markers. This maintains the nice property that deletion during iteration is perfectly safe.

**Comment by zminhquanz** [comment]
May that help to fix or improve perfomance ?

**Comment by hrydgard** [comment]
Just improve performance, and generally not by a lot. But every little bit helps.

File: Common/Hashmaps.h [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+	};
+	std::vector<Pair> map;
+	int capacity_;
+	int count_ = 0;
+};
+
+// Like the above, uses linear probing for cache-friendliness.
+// Does not perform hashing at all so expects well-distributed keys.
+template <class Value>
+class PrehashMap {
```
Comment (unknownbrackets) [inline Common/Hashmaps.h]: Hmm, I feel like we could have an abstract parent with 80% of the code, and have each of these be a thin subclass with some template params (i.e. KeyStorage template param or something.)

-[Unknown]

File: GPU/D3D11/DrawEngineD3D11.cpp [hunk 2]
```diff
@@ -111,18 +112,18 @@ void DrawEngineD3D11::InitDeviceObjects() {
 }
 
 void DrawEngineD3D11::ClearTrackedVertexArrays() {
-	for (auto &vai : vai_) {
-		delete vai.second;
-	}
-	vai_.clear();
+	vai_.Iterate([&](uint32_t hash, VertexArrayInfoD3D11 *vai){
```
Comment (unknownbrackets) [inline GPU/D3D11/DrawEngineD3D11.cpp]: I wonder if this is common enough we just have a specialized PointerDenseMap or something with a DeleteAll?  I suppose we have Release and other variants so maybe not...

-[Unknown]

Comment (hrydgard) [inline GPU/D3D11/DrawEngineD3D11.cpp]: Yeah, lots of things can be done :) A simple DeleteAll would take care of a bunch of these at least.

File: Common/Hashmaps.h (hunk 1 above)
Comment (hrydgard) [inline Common/Hashmaps.h]: Yeah, probably. Just keeping it simple to start with.

**Pull request merged**